### ISSUE TITLE
Slow bulkload performance due to high GC for couple of hundred million rows

### ISSUE THREAD

**Issue by puneetkp** [unicredit/hbase-rdd#15]
The library is good for writing smaller tables ~ couple of million rows of data. However, when we try inserting hundreds of millions of rows of data in HBase using it, we see poor performance.

From preliminary examination you can see that the GC time for the spark tasks increases as the data increases. 

We have done a couple of reductions in the object creation for 'toHBaseBulk'  and that has helped us reduce GC overhead. Please let me know if it seems of interest.

Thanks

**Comment by vidma** [comment]
P.S. We noticed that creating more regions/splits in hbase kinda minimizes the problem, but yes, it would be nice to have it faster.

**Comment by puneetkp** [comment]
You are right, increasing the regions and number of HFiles does reduce the GC impact. How do I share the suggested changes for review ?

**Comment by vidma** [comment]
fork github repo and Pull Request of course.

**Comment by vidma** [comment]
so could we see your changes @puneetkp

**Comment by fralken** [comment]
When you do a bulk load, you can tune the performance with:
- number of regions
- number of hfiles per region
You have (num of regions * num hfiles per region) tasks that execute the bulk load. The more tasks are executed, the less memory per task is used.

How to choose the number of regions? You can compute the total size of your data (roughly) and divide for the desired size of a region. Then you can use `computeSplits()` to get an array of split keys that you can use to create the table.

How to choose the number of HFiles? Probably the best option is to have 1 Hfile per region, but by default a region can have at most 32 HFiles, so you can increase this number in `toHBaseBulk` (e.g. `rdd.toHBaseBulk(table, cf, numFilesPerRegion=32)`).

E.g. if you have 10 regions and 32 HFiles per region you'll have 320 tasks in parallel, thus reducing memory per task and increasing parallelism.

**Comment by puneetkp** [comment]
I have forked the repo and created the new pull request.
The change involves avoiding creating the Map object for the column
mappings for each row. Also to avoid creating TreeSet objects.

**Comment by fralken** [comment]
We released version 0.6.0, with a new implementation of bulk write.
